FUSETOOLS2-1034 - tutorial registry updates

bfitzpat opened the pull request:
This will be a multiple commit project and this is the first chunk of work.

First commit adds:

* New setting to Automatically add the default Didact tutorials at startup if they are not present
* New command vscode.didact.registry.clear (not shown in the Didact Tutorials view, but present in the Command Palette)
* New command vscode.didact.registry.addJson used to add a new tutorial to the registry (JSON looks like ```{
			"name" : `tutorial name`,
			"category" : `exact category name`,
			"sourceUri" : `file/http URI to file`,
		};```)
* New tests validate these additional functions and verify that the older registerTutorialWithCategory method still works (though it now hands off to a registerTutorialWithArgs method)
* Added a class to handle the name/category/uri details for a new tutorial to more explicitly codify the format
* Removes the calls to clear the registry on startup and shutdown, relying on what's stored in the workspace settings

Second round of changes:
* New right-click menu `Didact: Register Didact Tutorial` on Didact files in the Explorer view enables user-registered tutorials to appear in `Didact Tutorials` view
* New right-click menu `Remove Didact Tutorial` on tutorials in `Didact Tutorials` view enables tutorial removal

Third round:
* Added an example of how to register a new tutorial via a Didact link along with all the other new bits

Fourth round:
* Updated the user prompts when registering a new tutorial. We now validate the name and provide an editable dropdown list of existing categories.

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

bfitzpat commented:
The markdown file examples/registry.example.didact.md walks through the various bits changed/added for this particular effort. The older method to register a tutorial still works, but there are other options now. 

Avoiding adding an "edit" for now but curious what you think of the changes.

bfitzpat commented:
Animated gif showing some of the functionality in action.
![Tutorial Registry Demo 23-MAR-2021](https://user-images.githubusercontent.com/530878/112202344-e53a5680-8bd6-11eb-9ddb-9300c94406a3.gif)

lhein reviewed (approved):
Looks good to me Brian. Nice addition!

apupier reviewed (changes requested):
nice addition

- using the "You can register it with the Didact Tutorials view with this link." in a workspace that contains space in path is registering a tutorial that fails to satrt
![errorWithSpaceInPath](https://user-images.githubusercontent.com/1105127/112282826-c3e57300-8c87-11eb-83da-9e95c2633b74.gif)
- when registering a tutorial with same name and category, it is silently failing. The best would be to have validation in the fields (will require to ask first for category and then for tutorial name). it can also be as first iteration to provide a notification mentioining that there is alaready a tutorial
- would be nice to better handle the case of hitting enter with no category provided. I thought that it would create the tutorial without category. In fact, it aborts the command. So either providing a "no category", or warn user that it is mandatory.

lhein commented on examples/registry.example.didact.md:
```diff
@@ -0,0 +1,33 @@
+
+## Adding via VS Code Extension Startup
+
+You can create your own VS Code extension to register a new tutorial with the Didact Tutorials view. And we have written a Didact tutorial to help you do just that!
+
+If you look at the Didact Tutorials view, you should be able to find the `Didact` category and beneath that find `Creating a New Didact Tutorial Extension`. That will step you through creating the new extension and your Didact tutorial file, then registering the new file with the view. 
+
+## *New* Adding via right-click `Didact: Register Didact Tutorial` menu in Explorer
```
good point

bfitzpat commented:
Thanks for the great feedback @apupier and @lhein ! I will work through it today

bfitzpat commented:
Addressed all the feedback and improved the user input 
![register-tutorial-demo](https://user-images.githubusercontent.com/530878/112500262-d88c3e80-8d4d-11eb-876b-582fd968d07d.gif)

bfitzpat commented:
Working through some oddness with the category selection box.

apupier reviewed (changes requested):
- validation is ignored when clicking outiside of the inputbox ![validationPassedIfClickingOutside](https://user-images.githubusercontent.com/1105127/112607375-6f71fd00-8e19-11eb-8ac4-4114ed997049.gif)

bfitzpat commented:
@apupier - added ignoreFocusOut to both the name field and category quickpick. Good catch!

bfitzpat commented:
@apupier or @lhein  It's odd though... you can only exit out by hitting Enter or Escape. If you Escape twice, you can get out. Any suggestions on how to make this better?

bfitzpat commented:
Found the fix. I was ignoring the undefined return too late in the process. It works as expected now (i.e. you can Escape out of either of the fields and it cancels you out immediately)

apupier commented on src/test/suite/registryWorkspace.test.ts:
```diff
@@ -0,0 +1,81 @@
+			await vscode.commands.executeCommand(START_DIDACT_COMMAND, didactFileUri);
+			try {
+				const predicate = () => didactManager.active() != undefined;
+				await waitUntil(predicate, { timeout: EDITOR_OPENED_TIMEOUT, intervalBetweenAttempts: 1000 }).then(() => {
+					expect(didactManager.active()?.getCurrentTitle()).to.equal("Local Didact Tutorial");
+				});
```
```suggestion
				await waitUntil(predicate, { timeout: EDITOR_OPENED_TIMEOUT, intervalBetweenAttempts: 1000 });
				expect(didactManager.active()?.getCurrentTitle()).to.equal("Local Didact Tutorial");
```

apupier commented on src/test/suite/registryWorkspace.test.ts:
```diff
@@ -0,0 +1,81 @@
+					expect(didactManager.active()?.getCurrentTitle()).to.equal("Local Didact Tutorial");
+				});
+			} catch (error) {
+				assert.fail(error);
+			}
+		} catch (error) {
+			assert.fail(error);
```
I think that either grouping hthe 2 try catch either providing more detaisl in fail assertion would be clearer

apupier commented on src/utils.ts:
```diff
@@ -250,3 +303,146 @@ export async function removeFilesAndFolders(workspacename: string, filesAndFolde
+}
+
+function validateTutorialNameInput(value: string, tutorialsForValidation : string[] ): string | null {
+	if (typeof value === "string") {
+	  if (value.trim().length === 0) {
+		return "Empty tutorial name is not allowed";
+	  }
+	  if (tutorialsForValidation && tutorialsForValidation.indexOf(value) > -1) {
+		  return "Tutorial with that name already exists. Tutorial names must be unique."
+	  }
+	  return null;
+	}
+  	return `${value} is invalid`;
```
mix of space and tabs

apupier commented on src/utils.ts:
```diff
@@ -250,3 +303,146 @@ export async function removeFilesAndFolders(workspacename: string, filesAndFolde
+				} else {
+					// include currently typed option if it isn't already there
+					if (categories.indexOf(quickPick.value.trim()) === -1) {
+                        quickPick.items = [{ label: quickPick.value.trim() }, ...options];
+                    }
```
space isntead of tabs

apupier commented on src/utils.ts:
```diff
@@ -250,3 +303,146 @@ export async function removeFilesAndFolders(workspacename: string, filesAndFolde
+			});
+		}
+		quickPick.onDidAccept(_ => {
+			if (quickPick.value.trim().length > 0 || selectedItems.length > 0 || quickPick.activeItems.length > 0) {
```
there is a trim here but not on the validation. it means that if a user add an existign name with space after, the validation is not triggered but the created value is erasing the previous template

bfitzpat replied:
@apupier  The trim is in both places... if you look at line 407, the trim is used there during the validation as well.

apupier replied:
ok, so do not know where is the glitch but the tutirial is created withotu space and validation does not catch extra space
![validationNotForbiddingSameNameExceptSpaceAtEnd](https://user-images.githubusercontent.com/1105127/112836858-6b551200-909b-11eb-821d-ffbcc2569375.gif)

bfitzpat replied:
found it. different spot (see the `validateTutorialNameInput` method)

apupier commented on src/utils.ts:
```diff
@@ -358,7 +358,10 @@ function validateTutorialNameInput(value: string, tutorialsForValidation : strin
 			return "Empty tutorial name is not allowed";
 		}
-		if (tutorialsForValidation && tutorialsForValidation.indexOf(value) > -1) {
+		if (value.endsWith(' ')) {
```
have you tried when starting with space? not tried but I think that it will needed to check for it too

bfitzpat replied:
It works. The additional trim in line 364 covers it. (Just tried locally)

bfitzpat replied:
I can add an additional validation check for it to be more explicit, but the trim resolves it even if the space is there (it just doesn't complain about it)

bfitzpat replied:
added

apupier commented on src/test/suite/registryWorkspace.test.ts:
```diff
@@ -0,0 +1,80 @@
+
+suite('Tutorial Registry Test Suite', () => {
+
+	const uriToRemoteDidactAdoc = 'https://raw.githubusercontent.com/redhat-developer/vscode-didact/master/demos/asciidoc/simple-example.didact.adoc';
```
unused variable
